### PR TITLE
363-Discrete Gaussian kernel 

### LINKED ISSUE #363: Discrete Gaussian kernel (gaussian_1d) should be adapted for discrete filtering
**Describe the bug**
The [current Gaussian kernel](https://github.com/Project-MONAI/MONAI/blob/5fcbf5ab16638670d890031825869874ed3d59ef/monai/networks/layers/convutils.py#L40-L59) simply samples the continuous Gaussian function:
```python
def gaussian_1d(sigma, truncated=4.0):
    """
    one dimensional gaussian kernel.

    Args:
        sigma: std of the kernel
        truncated: tail length

    Returns:
        1D numpy array
    """
    if sigma <= 0:
        raise ValueError("sigma must be positive")


    tail = int(sigma * truncated + 0.5)
    sigma2 = sigma * sigma
    x = np.arange(-tail, tail + 1)
    out = np.exp(-0.5 / sigma2 * x ** 2)
    out /= out.sum()
    return out
```

**Expected behavior**
Discrete Gaussian filtering should be compatible with scale-space theory and rely on modified Bessel function for its kernel weights. See e.g.:
https://en.wikipedia.org/wiki/Scale_space_implementation#The_discrete_Gaussian_kernel

**Additional context**
ITK implements such a behaviour in itk::GaussianOperator:
https://github.com/InsightSoftwareConsortium/ITK/blob/master/Modules/Core/Common/include/itkGaussianOperator.hxx

[Wikipedia](https://commons.wikimedia.org/wiki/File:Discrete_Gaussian_kernel.svg) also provides a basic python implementation (albeit relying on scipy and without rescaling):
```python
from scipy.special import iv
import numpy as np

def discrete_gaussian_kernel(n, t):
    T = np.exp(-t) * iv(n, t)
    return T
```

### PR DESCRIPTION
**Pull request by wyli**
Fixes #363.

### Description
adds a refined gaussian 1d kernel
with utilities of:
- separable filtering
- polyval

### Status
**Ready**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [x] New tests added to cover the changes.
- [x] Integration tests passed locally by running `./runtests.sh --codeformat --coverage`.
- [x] In-line docstrings updated.
- [x] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

**Comment by tvercaut** [comment]
Nice to see this taking shape! By skimming through it, I saw the call to `torch.no_grad`. Is the implementation not differentiable?

If not, what is preventing differentiability at the moment except things like calling `np.exp` rather than `torch.exp`?

In any case, expanding a bit the documentation to explain what "refined" means (i.e. pointing to https://en.wikipedia.org/wiki/Scale_space_implementation#The_discrete_Gaussian_kernel) and highlighting that the current "refined" implementation is not differentiable.

**Comment by wyli** [comment]
> Nice to see this taking shape! By skimming through it, I saw the call to `torch.no_grad`. Is the implementation not differentiable?
> 
> If not, what is preventing differentiability at the moment except things like calling `np.exp` rather than `torch.exp`?
> 
> In any case, expanding a bit the documentation to explain what "refined" means (i.e. pointing to https://en.wikipedia.org/wiki/Scale_space_implementation#The_discrete_Gaussian_kernel) and highlighting that the current "refined" implementation is not differentiable.

thanks @tvercaut! for the gradient, pytorch complains about some variable inplace operations has no gradient, I guess it's because of the recursions in the _modified_bessel_i loop, I'll take another look tomorrow.

File: monai/networks/layers/convutils.py [hunk 1]
```diff
@@ -97,7 +98,70 @@ def gaussian_1d(sigma: Union[float, torch.Tensor], truncated: float = 4.0) -> np
...
+        x = torch.arange(-tail, tail + 1).float()
+        t = 1 / (torch.tensor(2.0).sqrt() * sigma)
+        out = 0.5 * ((t * (x + 0.5)).erf() - (t * (x - 0.5)).erf())
+        return out.clamp(min=0)
+    if approx.lower() == "refined":
+        with torch.no_grad():
+            out = [_modified_bessel_0(sigma), _modified_bessel_1(sigma)]
+            idx = 2
+            while len(out) <= tail:
+                out.append(_modified_bessel_i(idx, sigma))
```
Comment (tvercaut) [inline monai/networks/layers/convutils.py:111]: I guess this might be the offending operation for differentiation purposes. Anyway, a simpler loop over an array of pre-allocated size might be better than an append approach (given that the size is known beforehand anyway).

File: monai/networks/layers/convutils.py [hunk 2]
```diff
@@ -78,13 +78,17 @@ def calculate_out_shape(
...
-def gaussian_1d(sigma: Union[float, torch.Tensor], truncated: float = 4.0) -> np.ndarray:
+def gaussian_1d(sigma: Union[float, torch.Tensor], truncated: float = 4.0, approx: str = "simple"):
     """
-    one dimensional gaussian kernel.
+    one dimensional Gaussian kernel.
 
     Args:
         sigma: std of the kernel
         truncated: tail length
+        approx: Discrete Gaussian kernel type, available options are "simple" and "refined".
```
Comment (tvercaut) [inline monai/networks/layers/convutils.py:88]: I would rather refer to these as "erf" (instead of "simple") and "scale-space" (instead of refined)

Comment (tvercaut) [reply monai/networks/layers/convutils.py:88]: It would also be good to have an even simpler "sampled approach", e.g.
```python
    if approx.lower() == "sampled":
        x = torch.arange(-tail, tail + 1).float()
        t = 1 / (torch.tensor(2.0*np.pi).sqrt() * sigma)
        out = t* torch.exp((-x.pow(2.0) / float(2 * sigma ** 2)))
        return out.clamp(min=0)
```

File: monai/networks/layers/convutils.py [hunk 3]
```diff
@@ -93,11 +97,74 @@ def gaussian_1d(sigma: Union[float, torch.Tensor], truncated: float = 4.0) -> np
...
-    t = 1 / (torch.tensor(2.0).sqrt() * sigma)
-    out = 0.5 * ((t * (x + 0.5)).erf() - (t * (x - 0.5)).erf())
-    return out.clamp(min=0)
+    if approx.lower() == "simple":
+        x = torch.arange(-tail, tail + 1, dtype=torch.float)
+        t = 1.0 / (torch.tensor(2.0).sqrt() * sigma)
+        out = 0.5 * ((t * (x + 0.5)).erf() - (t * (x - 0.5)).erf())
+        return out.clamp(min=0)
+    if approx.lower() == "refined":
+        out = [_modified_bessel_0(sigma), _modified_bessel_1(sigma)]
```
Comment (tvercaut) [inline monai/networks/layers/convutils.py:110]: Looks like there is a confusion between standard deviation and variance. Note that ITK asks for variance in the Gaussian operator: https://github.com/InsightSoftwareConsortium/ITK/blob/master/Modules/Core/Common/include/itkGaussianOperator.hxx

File: monai/networks/layers/convutils.py [hunk 4]
```diff
@@ -93,11 +97,74 @@ def gaussian_1d(sigma: Union[float, torch.Tensor], truncated: float = 4.0) -> np
...
+        while len(out) <= tail:
+            out.append(_modified_bessel_i(len(out), sigma))
```
Comment (tvercaut) [inline monai/networks/layers/convutils.py:112]: Would it not be better to pre-allocate the array and iterate over rather than append to it?

**Comment by tvercaut** [comment]
I played with the code and as illustrated with a simple visual test, there is a confusion betwen standard deviation and variance:
```python
%matplotlib inline
%config InlineBackend.figure_format = 'svg'
import matplotlib.pyplot as plt

extent = 6

xidx = range(-extent,extent)
print(xidx)
legend = []

for sigma in [0.5, 1.0, 2.0, 4.0]:
  print(f"sigma={sigma}")

  kernel_simple = gaussian_1d(sigma,truncated=extent/sigma,approx="simple").numpy()
  print(f"kernel_simple.sum()={kernel_simple.sum()}")

  kernel_simpler = gaussian_1d(sigma,truncated=extent/sigma,approx="simpler").numpy()
  print(f"kernel_simpler.sum()={kernel_simpler.sum()}")

  # Note the need to square sigma here
  kernel_refined = gaussian_1d(sigma**2,truncated=extent/sigma**2,approx="refined").numpy()
  print(f"kernel_refined.sum()={kernel_refined.sum()}")

  plt.plot(kernel_simple,'--')
  plt.plot(kernel_simpler,':')
  plt.plot(kernel_refined)
  legend.extend([f"simple s={sigma}",f"simpler s={sigma}",f"refined s={sigma}"])

plt.legend(legend)
```
This was meant to reproduce the graph seen in https://en.wikipedia.org/wiki/Scale_space_implementation#/media/File:Discrete_Gaussian_kernel.svg

Could be nice to turn this in a unit test.

File: monai/networks/layers/convutils.py [hunk 5]
```diff
@@ -93,11 +97,74 @@ def gaussian_1d(sigma: Union[float, torch.Tensor], truncated: float = 4.0) -> np
...
+        ans = out[:0:-1]
+        ans.extend(out)
+        ans = torch.stack(ans) * torch.exp(-sigma)
+        ans /= ans.sum()
```
Comment (tvercaut) [inline monai/networks/layers/convutils.py:116]: Also, I don't think we should normalise to sum up to one. If the truncation is large, it should indeed sum up to less than one. At least, maybe the normalisation to one should be an option available for all kernel implementations?

**Comment by wyli** [comment]
> I played with the code and as illustrated with a simple visual test, there is a confusion betwen standard deviation and variance:
> 
> ```python
> %matplotlib inline
> %config InlineBackend.figure_format = 'svg'
> import matplotlib.pyplot as plt
> 
> extent = 6
> 
> xidx = range(-extent,extent+1)
> legend = []
> 
> for sigma in [0.5, 1.0, 2.0, 4.0]:
>   print(f"sigma={sigma}")
> 
>   kernel_simple = gaussian_1d(sigma,truncated=extent/sigma,approx="simple").numpy()
>   print(f"kernel_simple.sum()={kernel_simple.sum()}")
> 
>   kernel_simpler = gaussian_1d(sigma,truncated=extent/sigma,approx="simpler").numpy()
>   print(f"kernel_simpler.sum()={kernel_simpler.sum()}")
> 
>   # Note the need to square sigma here
>   kernel_refined = gaussian_1d(sigma**2,truncated=extent/sigma**2,approx="refined").numpy()
>   print(f"kernel_refined.sum()={kernel_refined.sum()}")
> 
>   plt.plot(xidx,kernel_simple,'--')
>   plt.plot(xidx,kernel_simpler,':')
>   plt.plot(xidx,kernel_refined)
>   legend.extend([f"simple s={sigma}",f"simpler s={sigma}",f"refined s={sigma}"])
> 
> plt.legend(legend)
> ```
> 
> This was meant to reproduce the graph seen in https://en.wikipedia.org/wiki/Scale_space_implementation#/media/File:Discrete_Gaussian_kernel.svg
> 
> Could be nice to turn this in a unit test.

this is very helpful, thanks! (the wiki figure uses variance, here we use std. after fixing the notations, I'm able to reproduce it.)

**Comment by wyli** [comment]
Hi @Nic-Ma, all new features are tested and all of them are non-breaking changes (passed the integration tests locally).

**Comment by Nic-Ma** [comment]
Hi @tvercaut ,

As you had several comments and I am not very familiar with this Gaussian kernel topic, do you think it's OK to merge it now?
It looks good to me overall.

Thanks.

**Comment by tvercaut** [comment]
@Nic-Ma @wyli : I haven't had the time to test the new version but reading through the code it seems fine to me. The fact that @wyli could replicate the wikipedia plot is also good evidence that the code does what it's suppose to do.

**Review by tvercaut — APPROVED** [review]
